lib/fibonacci.rb

```diff
@@ -1,9 +1,84 @@
 # Improved Fibonacci
 
-# Time Complexity - ?
-# Space Complexity - ?  (should be O(n))
+# Time Complexity -  I'm pretty sure this is O(n) time complexity. First off, there is only one recursive
+#                    call per recursive step, which prevents that 2^n branching we see in the non-memoized
+#                    Fib function, and secondly the recursives call directly respond to the length of the stack,
+#                    which is size n for the nth Fibonacci number.
+# Space Complexity - O(n) - the array use to store the previous values will grow along with the stack --
+#                    that is to say, the nth Fibonacci number will have an n - level stack and an array of
+#                    of length n to hold all previous Fibonacci numbers. n + n = O(2n) = 2*O(n) = O(n) space
+#                    complexity.
 # Hint, you may want a recursive helper method
 # Hint:  Fib(0) = 0, Fib(1) = 1, work up from there
 def fibonacci(n)
```

Comment:
👍 This works, but can you solve it without having an array of n size?

Reply:
Ooh, I was able to find a solution with an array of size 2, but I think that could be easily rewritten as simply two parameters holding the most recent two values in the recursive function call. There's also another version that literally only takes one or two lines of code (and is memoized) that I can't think of right now.